keps/sig-scheduling/2458-preferred-fit-strategy/README.md

```diff
@@ -0,0 +1,664 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Define a new `PreferredFitStrategy` type as follows:
+
+```go
+
+type StrategyType string
+
+const (
+    LeastAllocated StrategyType = "LeastAllocated"
+    MostAllocated StrategyType = "MostAllocated"
+    BalancedAllocated StrategyType = "BalancedAllocated"
+    RequestedToCapacityRatio StrategyType = "RequestedToCapacityRatio"
+)
+
+type PreferredFitStrategy struct {
+	metav1.TypeMeta
+
+    // Strategy selects which strategy to run.
+    Strategy StrategyType
+    
+    // Resources to consider when scoring.
+    // The default resource set includes "cpu" and "memory" with an equal weight.
+	// Allowed weights go from 1 to 100.
+    Resources []ResourceSpec
+
+    // Arguemnts specific to RequestedToCapacityRatio strategy.
+    RequestedToCapacityRatio *RequestedToCapacityRatio
+}
+
+type RequestedToCapacityRatio struct {
+    // Points defining priority function shape
+    Shape []UtilizationShapePoint
+}
+
+// Note that the two types defined below already exist in the scheduler's component config API.
+// ResourceSpec represents a single resource.
+type ResourceSpec struct {
+	// Name of the resource.
+	Name string
+	// Weight of the resource.
+	Weight int64
+}
+
+// UtilizationShapePoint represents a single point of a priority function shape.
+type UtilizationShapePoint struct {
+	// Utilization (x axis). Valid values are 0 to 100. Fully utilized node maps to 100.
+	Utilization int32
+	// Score assigned to a given utilization (y axis). Valid values are 0 to 10.
+	Score int32
+}
+
+```
+
+Add `PreferredFitStrategy` to `NodeResourcesFitArgs`:
+
+```go
+
+// NodeResourcesFitArgs holds arguments used to configure the NodeResourcesFit plugin.
+type NodeResourcesFitArgs struct {
+	metav1.TypeMeta
+
+	// IgnoredResources is the list of resources that NodeResources fit filter
+	// should ignore.
+	IgnoredResources []string
+	// IgnoredResourceGroups defines the list of resource groups that NodeResources fit 
+    // filter should ignore.
+	// e.g. if group is ["example.com"], it will ignore all resource names that begin
+	// with "example.com", such as "example.com/aaa" and "example.com/bbb".
+	// A resource group name can't contain '/'.
+	IgnoredResourceGroups []string
+    
+    // PreferredFitStrategy selects the node resource scoring strategy.
+    PreferredFitStrategy *PreferredFitStrategy
+}
+```
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+- Unit and integration tests covering core changes.
```

Comment:
Would we duplicate the tests for the existing plugins? Worth mentioning.

Reply:
I clarified the plan is to test the new configuration path. Ideally the logic and its unit tests will stay where they are since we need to keep the old plugins for now.

Reply:
but we would duplicate the test from the 3 plugins into the Fit plugin, right?

Reply:
I think we can avoid that by a proper code refactor. The most important thing is that all code paths have test coverage.

Reply:
sg